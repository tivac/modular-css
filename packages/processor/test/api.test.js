"use strict";

const postcss  = require("postcss");
const namer    = require("@modular-css/test-utils/namer.js");
const relative = require("@modular-css/test-utils/relative.js");

const Processor = require("../processor.js");

describe("/processor.js", () => {
    describe("API", () => {
        let processor;

        beforeEach(() => {
            processor = new Processor({
                namer,
            });
        });

        it("should be a function", () => {
            expect(typeof Processor).toBe("function");
        });

        describe(".root()", () => {
            it("should process a postcss Root", async () => {
                const file = "./simple.css";
                const root = postcss.parse(".wooga { }", { from : file });
                const { exports : compositions, details } = await processor.root(file, root);

                expect(compositions).toMatchSnapshot();
                expect(details.classes).toMatchSnapshot();
                expect(details.processed.root.toResult().css).toMatchSnapshot();
            });
        });

        describe(".string()", () => {
            it("should process a string", async () => {
                const { exports : compositions, details } = await processor.string("./simple.css", ".wooga { }");

                expect(compositions).toMatchSnapshot();
                expect(details.classes).toMatchSnapshot();
                expect(details.processed.root.toResult().css).toMatchSnapshot();
            });
        });

        describe(".file()", () => {
            it("should process a relative file", async () => {
                const { exports : compositions, details } = await processor.file(
                    "./packages/processor/test/specimens/simple.css"
                );

                expect(compositions).toMatchSnapshot();
                expect(details.classes).toMatchSnapshot();
                expect(details.processed.root.toResult().css).toMatchSnapshot();
            });

            it("should process an absolute file", async () => {
                const { exports : compositions, details } = await processor.file(
                    require.resolve("./specimens/simple.css")
                );

                expect(compositions).toMatchSnapshot();
                expect(details.classes).toMatchSnapshot();
                expect(details.processed.root.toResult().css).toMatchSnapshot();
            });

            it("should wait for dependencies to be processed before composing", async () => {
                const results = await Promise.all([
                    processor.file(require.resolve("./specimens/overlapping/entry1.css")),
                    processor.file(require.resolve("./specimens/overlapping/entry2.css")),
                ]);

                expect(results.map((result) => result.exports)).toMatchSnapshot();
            });
        });

        describe(".has()", () => {
            it("should return a boolean", async () => {
                await processor.string(
                    "./simple.css",
                    ".wooga { }"
                );

                expect(processor.has("./simple.css")).toBe(true);
                expect(processor.has("./nope.css")).toBe(false);
            });

            it("should normalize inputs before checking for existence", async () => {
                await processor.string(
                    "./simple.css",
                    ".wooga { }"
                );

                expect(processor.has("../modular-css/simple.css")).toBe(true);
            });
        });

        describe(".normalize()", () => {
            it("should normalize inputs", async () => {
                expect(relative([ processor.normalize("../modular-css/simple.css") ])).toMatchSnapshot();
            });
        });

        describe(".remove()", () => {
            it("should remove a relative file", async () => {
                await processor.string(
                    "./simple.css",
                    ".wooga { }"
                );

                processor.remove("./simple.css");

                expect(relative(processor.dependencies())).toMatchSnapshot();
            });

            it("should remove an absolute file", async () => {
                await processor.string(
                    "./packages/processor/test/specimens/simple.css",
                    ".wooga { }"
                );

                processor.remove(require.resolve("./specimens/simple.css"));

                expect(relative(processor.dependencies())).toMatchSnapshot();
            });

            it("should remove multiple files", async () => {
                await processor.string("./a.css", ".a { }");
                await processor.string("./b.css", ".b { }");
                await processor.string("./c.css", ".c { }");

                processor.remove([
                    "./a.css",
                    "./b.css",
                ]);

                expect(relative(processor.dependencies())).toMatchSnapshot();
            });

            it("should return an array of removed files", async () => {
                await processor.string("./a.css", ".a { }");
                await processor.string("./b.css", ".b { }");
                await processor.string("./c.css", ".c { }");

                expect(
                    relative(
                        processor.remove([
                            "./a.css",
                            "./b.css",
                        ])
                    )
                ).toMatchSnapshot();
            });
        });

        describe(".invalidate()", () => {
            const status = (source) =>
                Object.entries(source).map(([ key, value ]) =>
                    ([ relative([ key ])[0], value.valid ])
                );

            it("should invalidate a relative file", async () => {
                await processor.string(
                    "./simple.css",
                    ".wooga { }"
                );

                processor.invalidate("./simple.css");

                expect(status(processor.files)).toMatchSnapshot();
            });

            it("should invalidate an absolute file", async () => {
                await processor.string(
                    "./packages/processor/test/specimens/simple.css",
                    ".wooga { }"
                );

                processor.invalidate(require.resolve("./specimens/simple.css"));

                expect(status(processor.files)).toMatchSnapshot();
            });

            it("should throw if no file is passed", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");

                expect(() => processor.invalidate()).toThrowErrorMatchingSnapshot();
            });

            it("should throw if an invalid file is passed", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");

                expect(() => processor.invalidate("nope.css")).toThrow(/Unknown file: .+\bnope.css/);
            });

            it("should invalidate all dependents as well", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");

                processor.invalidate("./packages/processor/test/specimens/folder/folder.css");

                expect(status(processor.files)).toMatchSnapshot();
            });

            it("should reprocess invalidated files", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");

                processor.invalidate("./packages/processor/test/specimens/start.css");

                await processor.file("./packages/processor/test/specimens/start.css");

                expect(status(processor.files)).toMatchSnapshot();
            });
        });

        describe(".dependencies()", () => {
            it("should return the dependencies of the specified file", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");

                expect(
                    relative(processor.dependencies(require.resolve("./specimens/start.css")))
                )
                .toMatchSnapshot();
            });

            it("should return the overall order of dependencies if no file is specified", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");

                expect(relative(processor.dependencies())).toMatchSnapshot();
            });
        });

        describe(".output()", () => {
            it("should reject unknown files", async () => {
                await expect(processor.output({
                    files : [
                        "./not/a/real/file",
                    ],
                })).rejects.toThrow("Unknown file requested");
            });

            it("should return a postcss result", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");

                const result = await processor.output();

                expect(result.css).toMatchSnapshot();
            });

            it("should generate css representing the output from all added files", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");
                await processor.file("./packages/processor/test/specimens/simple.css");

                const result = await processor.output();

                expect(result.css).toMatchSnapshot();
            });

            it("should avoid duplicating files in the output", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");
                await processor.file("./packages/processor/test/specimens/local.css");

                const result = await processor.output();

                expect(result.css).toMatchSnapshot();
            });

            it("should generate a JSON structure of all the compositions", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");

                const result = await processor.output();

                expect(result.compositions).toMatchSnapshot();
            });

            it("should order output by dependencies, then alphabetically", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");
                await processor.file("./packages/processor/test/specimens/local.css");
                await processor.file("./packages/processor/test/specimens/composes.css");
                await processor.file("./packages/processor/test/specimens/deep.css");

                const result = await processor.output();

                expect(result.css).toMatchSnapshot();
            });

            it("should support returning output for specified relative files", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");
                await processor.file("./packages/processor/test/specimens/local.css");

                const result = await processor.output({
                    files : [
                        "./packages/processor/test/specimens/start.css",
                    ],
                });

                expect(result.css).toMatchSnapshot();
            });

            it("should support returning output for specified absolute files", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");
                await processor.file("./packages/processor/test/specimens/local.css");

                const result = await processor.output({
                    files : [
                        require.resolve("./specimens/start.css"),
                    ],
                });

                expect(result.css).toMatchSnapshot();
            });

            it("should allow for seperate source map output", async () => {
                await processor.file("./packages/processor/test/specimens/start.css");

                const result = await processor.output({
                    map : {
                        inline : false,
                    },
                });

                expect(result.map).toMatchSnapshot();
            });
        });

        describe(".compositions", () => {
            it("should return compositions for loaded files", async () => {
                await processor.file(require.resolve("./specimens/start.css"));

                expect(processor.compositions).toMatchSnapshot();
            });
        });

        describe(".resolve()", () => {
            it("should run resolvers until a match is found", () => {
                let ran = false;

                processor = new Processor({
                    resolvers : [
                        () => {
                            ran = true;
                        },
                    ],
                });

                expect(
                    relative([
                        processor.resolve(
                            require.resolve("./specimens/start.css"),
                            "./local.css"
                        ),
                    ])
                )
                .toMatchSnapshot();

                expect(ran).toBeTruthy();
            });

            it("should fall back to a default resolver", () => {
                processor = new Processor({
                    resolvers : [
                        () => undefined,
                    ],
                });

                expect(
                    relative([
                        processor.resolve(
                            require.resolve("./specimens/start.css"),
                            "./local.css"
                        ),
                    ])
                )
                .toMatchSnapshot();
            });
        });
    });
});
