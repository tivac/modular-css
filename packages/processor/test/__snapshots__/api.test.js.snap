// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/processor.js API .compositions should return compositions for loaded files 1`] = `Promise {}`;

exports[`/processor.js API .dependencies() should return the dependencies of the specified file 1`] = `
Array [
  "packages/processor/test/specimens/folder/folder.css",
  "packages/processor/test/specimens/local.css",
]
`;

exports[`/processor.js API .dependencies() should return the overall order of dependencies if no file is specified 1`] = `
Array [
  "packages/processor/test/specimens/folder/folder.css",
  "packages/processor/test/specimens/local.css",
  "packages/processor/test/specimens/start.css",
]
`;

exports[`/processor.js API .file() should process a relative file 1`] = `
Object {
  "wooga": Array [
    "mc_wooga",
  ],
}
`;

exports[`/processor.js API .file() should process a relative file 2`] = `
Object {
  "wooga": Array [
    "mc_wooga",
  ],
}
`;

exports[`/processor.js API .file() should process a relative file 3`] = `
".mc_wooga { color: red; }
"
`;

exports[`/processor.js API .file() should process an absolute file 1`] = `
Object {
  "wooga": Array [
    "mc_wooga",
  ],
}
`;

exports[`/processor.js API .file() should process an absolute file 2`] = `
Object {
  "wooga": Array [
    "mc_wooga",
  ],
}
`;

exports[`/processor.js API .file() should process an absolute file 3`] = `
".mc_wooga { color: red; }
"
`;

exports[`/processor.js API .file() should wait for dependencies to be processed before composing 1`] = `
Array [
  Object {
    "one": Array [
      "mc_three",
      "mc_two",
      "mc_one",
      "mc_one",
    ],
  },
  Object {
    "two": Array [
      "mc_three",
      "mc_two",
      "mc_one",
      "mc_two",
    ],
  },
]
`;

exports[`/processor.js API .invalidate() should invalidate a relative file 1`] = `
Array [
  Array [
    "simple.css",
    false,
  ],
]
`;

exports[`/processor.js API .invalidate() should invalidate all dependents as well 1`] = `
Array [
  Array [
    "packages/processor/test/specimens/start.css",
    false,
  ],
  Array [
    "packages/processor/test/specimens/local.css",
    false,
  ],
  Array [
    "packages/processor/test/specimens/folder/folder.css",
    false,
  ],
]
`;

exports[`/processor.js API .invalidate() should invalidate an absolute file 1`] = `
Array [
  Array [
    "packages/processor/test/specimens/simple.css",
    false,
  ],
]
`;

exports[`/processor.js API .invalidate() should reprocess invalidated files 1`] = `
Array [
  Array [
    "packages/processor/test/specimens/start.css",
    true,
  ],
  Array [
    "packages/processor/test/specimens/local.css",
    true,
  ],
  Array [
    "packages/processor/test/specimens/folder/folder.css",
    true,
  ],
]
`;

exports[`/processor.js API .invalidate() should throw if no file is passed 1`] = `"invalidate() requires a file argument"`;

exports[`/processor.js API .normalize() should normalize inputs 1`] = `
Array [
  "simple.css",
]
`;

exports[`/processor.js API .output() should allow for seperate source map output 1`] = `
Object {
  "file": "to.css",
  "mappings": "AAAA,wDAAA;AACA,aAAU,WAAW,EAAE;ACDvB,gDAAA;AAIA,YAAS,iBAAiB,EAAE;ACJ5B,gDAAA;AAEA,YDFA,UAAe,EACf,gBAAgB,ECCsB;AACtC,YFHA,uBAAoB,EEGe",
  "names": Array [],
  "sources": Array [
    "packages/processor/test/specimens/folder/folder.css",
    "packages/processor/test/specimens/local.css",
    "packages/processor/test/specimens/start.css",
  ],
  "sourcesContent": Array [
    "@value folder: white;
.folder { margin: 2px; }
",
    "@value one: red;
@value two: blue;
@value folder from \\"./folder/folder.css\\";

.booga { background: green; }
.looga { composes: booga; }
",
    "@value one, two, folder from \\"./local.css\\";
.wooga { composes: booga from \\"./local.css\\"; }
.booga { color: one; background: two; }
.tooga { border: 1px solid folder; }
",
  ],
  "version": 3,
}
`;

exports[`/processor.js API .output() should avoid duplicating files in the output 1`] = `
"/* packages/processor/test/specimens/folder/folder.css */
.mc_folder { margin: 2px; }
/* packages/processor/test/specimens/local.css */
.mc_booga { background: green; }
/* packages/processor/test/specimens/start.css */
.mc_booga { color: red; background: blue; }
.mc_tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should generate a JSON structure of all the compositions 1`] = `
Object {
  "packages/processor/test/specimens/folder/folder.css": Object {
    "folder": Array [
      "mc_folder",
    ],
  },
  "packages/processor/test/specimens/local.css": Object {
    "booga": Array [
      "mc_booga",
    ],
    "looga": Array [
      "mc_booga",
      "mc_looga",
    ],
  },
  "packages/processor/test/specimens/start.css": Object {
    "booga": Array [
      "mc_booga",
    ],
    "tooga": Array [
      "mc_tooga",
    ],
    "wooga": Array [
      "mc_booga",
      "mc_wooga",
    ],
  },
}
`;

exports[`/processor.js API .output() should generate css representing the output from all added files 1`] = `
"/* packages/processor/test/specimens/folder/folder.css */
.mc_folder { margin: 2px; }
/* packages/processor/test/specimens/simple.css */
.mc_wooga { color: red; }
/* packages/processor/test/specimens/local.css */
.mc_booga { background: green; }
/* packages/processor/test/specimens/start.css */
.mc_booga { color: red; background: blue; }
.mc_tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should order output by dependencies, then alphabetically 1`] = `
"/* packages/processor/test/specimens/folder/folder.css */
.mc_folder { margin: 2px; }
/* packages/processor/test/specimens/folder/folder2.css */
.mc_folder2 { color: green; }
/* packages/processor/test/specimens/folder/subfolder/subfolder.css */
.mc_subfolder { color: yellow; }
/* packages/processor/test/specimens/local.css */
.mc_booga { background: green; }
/* packages/processor/test/specimens/composes.css */
/* packages/processor/test/specimens/deep.css */
.mc_deep {
    color: black;
}
/* packages/processor/test/specimens/start.css */
.mc_booga { color: red; background: blue; }
.mc_tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should return a postcss result 1`] = `
"/* packages/processor/test/specimens/folder/folder.css */
.mc_folder { margin: 2px; }
/* packages/processor/test/specimens/local.css */
.mc_booga { background: green; }
/* packages/processor/test/specimens/start.css */
.mc_booga { color: red; background: blue; }
.mc_tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should support returning output for specified absolute files 1`] = `
"/* packages/processor/test/specimens/start.css */
.mc_booga { color: red; background: blue; }
.mc_tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should support returning output for specified relative files 1`] = `
"/* packages/processor/test/specimens/start.css */
.mc_booga { color: red; background: blue; }
.mc_tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .remove() should remove a relative file 1`] = `Array []`;

exports[`/processor.js API .remove() should remove an absolute file 1`] = `Array []`;

exports[`/processor.js API .remove() should remove multiple files 1`] = `
Array [
  "c.css",
]
`;

exports[`/processor.js API .remove() should return an array of removed files 1`] = `
Array [
  "a.css",
  "b.css",
]
`;

exports[`/processor.js API .resolve() should fall back to a default resolver 1`] = `
Array [
  "packages/processor/test/specimens/local.css",
]
`;

exports[`/processor.js API .resolve() should run resolvers until a match is found 1`] = `
Array [
  "packages/processor/test/specimens/local.css",
]
`;

exports[`/processor.js API .root() should process a postcss Root 1`] = `
Object {
  "wooga": Array [
    "mc_wooga",
  ],
}
`;

exports[`/processor.js API .root() should process a postcss Root 2`] = `
Object {
  "wooga": Array [
    "mc_wooga",
  ],
}
`;

exports[`/processor.js API .root() should process a postcss Root 3`] = `".mc_wooga { }"`;

exports[`/processor.js API .string() should process a string 1`] = `
Object {
  "wooga": Array [
    "mc_wooga",
  ],
}
`;

exports[`/processor.js API .string() should process a string 2`] = `
Object {
  "wooga": Array [
    "mc_wooga",
  ],
}
`;

exports[`/processor.js API .string() should process a string 3`] = `".mc_wooga { }"`;
